File: src/pool/job.ts
import crypto from 'crypto';
import { BlockTemplate } from '../bitcoin/rpc';
import { buildCoinbaseTransaction } from './coinbase';
import { sha256d, reverseBuffer } from '../crypto/sha256';
import { config } from '../config';

export interface MiningJob {
  jobId: string;
  blockHeight: number;
  coinbaseValue: number;
  prevHashStratum: string;
  prevHashRaw: string;
  versionHex: string;
  versionMaskHex: string;
  nBitsHex: string;
  nTimeHex: string;
  merkleBranchHex: string[];
  txsData: string[];
  defaultWitnessCommitment?: string;
  targetHex?: string;
  coinb1: string;
  coinb2: string;
  createdTime: number;
}

export class JobManager {
  private currentJob: MiningJob | null = null;
  private jobMap: Map<string, MiningJob> = new Map();
  private jobCounter = 0;

  /**
   * Format prevhash into standard Stratum 8x32-bit word Little-Endian format
   */
  private formatPrevHashForStratum(prevHashHex: string): string {
    const b = Buffer.from(prevHashHex, 'hex');
    const bLE = reverseBuffer(b);
    for (let i = 0; i < bLE.length; i += 4) {
      const t0 = bLE[i];
      const t1 = bLE[i + 1];
      bLE[i] = bLE[i + 3];
      bLE[i + 1] = bLE[i + 2];
      bLE[i + 2] = t1;
      bLE[i + 3] = t0;
    }
    return bLE.toString('hex');
  }

  /**
   * Calculate Merkle Tree Branch hashes for Stratum (for Coinbase at index 0)
   */
  private calculateMerkleBranch(txs: any[]): string[] {
    if (!txs || txs.length === 0) return [];

    let hashes: Buffer[] = txs.map((tx) => {
      const hashStr = typeof tx === 'string' ? tx : (tx.hash || tx.txid);
      if (hashStr) {
        return reverseBuffer(Buffer.from(hashStr, 'hex'));
      }
      const rawTx = Buffer.from(tx.data, 'hex');
      return sha256d(rawTx);
    });

    const branch: string[] = [];

    while (hashes.length > 0) {
      branch.push(hashes[0].toString('hex'));

      if (hashes.length === 1) break;

      const nextLevel: Buffer[] = [];
      for (let i = 1; i < hashes.length; i += 2) {
        const left = hashes[i];
        const right = i + 1 < hashes.length ? hashes[i + 1] : hashes[i];
        nextLevel.push(sha256d(Buffer.concat([left, right])));
      }
      hashes = nextLevel;
    }

    return branch;
  }

  /**
   * Calculate allowed AsicBoost version mask from block template vbavailable
   */
  private calculateVersionMask(template: BlockTemplate): string {
    // Default overt AsicBoost version-rolling mask: 0x1fffe000 (bits 13 to 28)
    let maskNum = 0x1fffe000;
    if (template.vbavailable) {
      // Respect bits offered by node
      let availableBits = 0;
      for (const key of Object.keys(template.vbavailable)) {
        const bit = template.vbavailable[key];
        availableBits |= 1 << bit;
      }
      if (availableBits > 0) {
        maskNum = availableBits | 0x1fffe000;
      }
    }
    return maskNum.toString(16).padStart(8, '0');
  }

  /**
   * Create new job from Bitcoin RPC BlockTemplate
   */
  public createJob(template: BlockTemplate): MiningJob {
    this.jobCounter++;
    const jobId = this.jobCounter.toString(16).padStart(4, '0');

    const prevHashStratum = this.formatPrevHashForStratum(template.previousblockhash);
    const merkleBranchHex = this.calculateMerkleBranch(template.transactions);
    const versionHex = template.version.toString(16).padStart(8, '0');
    const versionMaskHex = this.calculateVersionMask(template);
    const nBitsHex = template.bits;
    const nTimeHex = template.curtime.toString(16).padStart(8, '0');

    const coinbase = buildCoinbaseTransaction({
      blockHeight: template.height,
      coinbaseValue: template.coinbasevalue,
      minerAddress: config.walletAddress || config.poolFeeAddress || 'AWPuDcCymof8BRF9cfkxnLqmhn7ZPVPjEr',
      extranonce1Size: 4,
      extranonce2Size: 4,
      defaultWitnessCommitment: template.default_witness_commitment,
    });

    const job: MiningJob = {
      jobId,
      blockHeight: template.height,
      coinbaseValue: template.coinbasevalue,
      prevHashStratum,
      prevHashRaw: template.previousblockhash,
      versionHex,
      versionMaskHex,
      nBitsHex,
      nTimeHex,
      merkleBranchHex,
      txsData: template.transactions.map((t) => t.data),
      defaultWitnessCommitment: template.default_witness_commitment,
      targetHex: template.target,
      coinb1: coinbase.coinb1,
      coinb2: coinbase.coinb2,
      createdTime: Date.now(),
    };

    this.currentJob = job;
    this.jobMap.set(jobId, job);

    // Keep map of recent jobs (max 200 jobs for seamless job transitions)
    if (this.jobMap.size > 200) {
      const oldestKey = this.jobMap.keys().next().value;
      if (oldestKey) this.jobMap.delete(oldestKey);
    }

    return job;
  }

  public getCurrentJob(): MiningJob | null {
    return this.currentJob;
  }

  public getJob(jobId: string): MiningJob | undefined {
    return this.jobMap.get(jobId);
  }
}

File: pool/job.go
package pool

import (
	"encoding/hex"
	"fmt"
	"strconv"
	"sync"
	"time"

	"zpoolproxy/config"
	"zpoolproxy/crypto"
)

type MiningJob struct {
	JobId                    string   `json:"jobId"`
	BlockHeight              int64    `json:"blockHeight"`
	CoinbaseValue            int64    `json:"coinbaseValue"`
	PrevHashStratum          string   `json:"prevHashStratum"`
	PrevHashRaw              string   `json:"prevHashRaw"`
	VersionHex               string   `json:"versionHex"`
	VersionMaskHex           string   `json:"versionMaskHex"`
	NBitsHex                 string   `json:"nBitsHex"`
	NTimeHex                 string   `json:"nTimeHex"`
	MerkleBranchHex          []string `json:"merkleBranchHex"`
	TxsData                  []string `json:"txsData"`
	DefaultWitnessCommitment string   `json:"defaultWitnessCommitment,omitempty"`
	TargetHex                string   `json:"targetHex,omitempty"`
	Coinb1                   string   `json:"coinb1"`
	Coinb2                   string   `json:"coinb2"`
	CreatedTime              int64    `json:"createdTime"`
}

type JobManager struct {
	mu         sync.RWMutex
	cfg        *config.Config
	currentJob *MiningJob
	jobMap     map[string]*MiningJob
	jobCounter int64
}

func NewJobManager(cfg *config.Config) *JobManager {
	return &JobManager{
		cfg:    cfg,
		jobMap: make(map[string]*MiningJob),
	}
}

func (jm *JobManager) FormatPrevHashForStratum(prevHashHex string) string {
	b, err := hex.DecodeString(prevHashHex)
	if err != nil {
		return prevHashHex
	}
	bLE := crypto.ReverseBytes(b)
	for i := 0; i < len(bLE); i += 4 {
		t0 := bLE[i]
		t1 := bLE[i+1]
		bLE[i] = bLE[i+3]
		bLE[i+1] = bLE[i+2]
		bLE[i+2] = t1
		bLE[i+3] = t0
	}
	return hex.EncodeToString(bLE)
}

func (jm *JobManager) CalculateMerkleBranch(txs []map[string]interface{}) []string {
	if len(txs) == 0 {
		return []string{}
	}

	var hashes [][]byte
	for _, tx := range txs {
		if hashStr, ok := tx["hash"].(string); ok && hashStr != "" {
			b, err := hex.DecodeString(hashStr)
			if err == nil {
				hashes = append(hashes, crypto.ReverseBytes(b))
				continue
			}
		} else if txidStr, ok := tx["txid"].(string); ok && txidStr != "" {
			b, err := hex.DecodeString(txidStr)
			if err == nil {
				hashes = append(hashes, crypto.ReverseBytes(b))
				continue
			}
		} else if dataStr, ok := tx["data"].(string); ok && dataStr != "" {
			b, err := hex.DecodeString(dataStr)
			if err == nil {
				hashes = append(hashes, crypto.Sha256d(b))
				continue
			}
		}
	}

	var branch []string

	for len(hashes) > 0 {
		branch = append(branch, hex.EncodeToString(hashes[0]))

		if len(hashes) == 1 {
			break
		}

		// Correct Merkle Branch slicing fix: slice out index 0 (which was added to branch)
		remaining := hashes[1:]
		var nextLevel [][]byte
		for i := 0; i < len(remaining); i += 2 {
			left := remaining[i]
			right := left
			if i+1 < len(remaining) {
				right = remaining[i+1]
			}
			combined := append(left, right...)
			nextLevel = append(nextLevel, crypto.Sha256d(combined))
		}
		hashes = nextLevel
	}

	return branch
}

func (jm *JobManager) CreateJob(template map[string]interface{}) *MiningJob {
	jm.mu.Lock()
	defer jm.mu.Unlock()

	jm.jobCounter++
	jobId := fmt.Sprintf("%04x", jm.jobCounter)

	height, _ := template["height"].(float64)
	coinbaseval, _ := template["coinbasevalue"].(float64)
	prevblockhash, _ := template["previousblockhash"].(string)
	version, _ := template["version"].(float64)
	bits, _ := template["bits"].(string)
	curtime, _ := template["curtime"].(float64)
	target, _ := template["target"].(string)
	witnessCommitment, _ := template["default_witness_commitment"].(string)

	var txsData []string
	var rawTxs []map[string]interface{}

	if txList, ok := template["transactions"].([]interface{}); ok {
		for _, item := range txList {
			if txMap, ok := item.(map[string]interface{}); ok {
				rawTxs = append(rawTxs, txMap)
				if data, ok := txMap["data"].(string); ok {
					txsData = append(txsData, data)
				}
			}
		}
	}

	prevHashStratum := jm.FormatPrevHashForStratum(prevblockhash)
	merkleBranchHex := jm.CalculateMerkleBranch(rawTxs)
	versionHex := fmt.Sprintf("%08x", uint32(version))
	nBitsHex := bits
	nTimeHex := fmt.Sprintf("%08x", uint32(curtime))

	minerAddr := jm.cfg.WalletAddress
	if minerAddr == "" {
		minerAddr = jm.cfg.PoolFeeAddress
	}

	cbParts := BuildCoinbaseTransaction(
		jm.cfg,
		int64(height),
		int64(coinbaseval),
		minerAddr,
		4,
		4,
		witnessCommitment,
	)

	job := &MiningJob{
		JobId:                    jobId,
		BlockHeight:              int64(height),
		CoinbaseValue:            int64(coinbaseval),
		PrevHashStratum:          prevHashStratum,
		PrevHashRaw:              prevblockhash,
		VersionHex:               versionHex,
		VersionMaskHex:           "1fffe000",
		NBitsHex:                 nBitsHex,
		NTimeHex:                 nTimeHex,
		MerkleBranchHex:          merkleBranchHex,
		TxsData:                  txsData,
		DefaultWitnessCommitment: witnessCommitment,
		TargetHex:                target,
		Coinb1:                   cbParts.Coinb1,
		Coinb2:                   cbParts.Coinb2,
		CreatedTime:              time.Now().UnixMilli(),
	}

	jm.currentJob = job
	jm.jobMap[jobId] = job

	for len(jm.jobMap) > 200 {
		oldestID := ""
		oldestCounter := int64(1<<63 - 1)

		for k := range jm.jobMap {
			if n, err := strconv.ParseInt(k, 16, 64); err == nil {
				if n < oldestCounter {
					oldestCounter = n
					oldestID = k
				}
			}
		}

		if oldestID == "" {
			break
		}
		delete(jm.jobMap, oldestID)
	}

	return job
}

func (jm *JobManager) GetCurrentJob() *MiningJob {
	jm.mu.RLock()
	defer jm.mu.RUnlock()
	return jm.currentJob
}

func (jm *JobManager) GetJob(jobId string) *MiningJob {
	jm.mu.RLock()
	defer jm.mu.RUnlock()
	return jm.jobMap[jobId]
}
